File: q1027/q1027.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    int longestArithSeqLength(vector<int> &nums)
    {
        int maxLen = 0;
        int n = nums.size();
        vector<unordered_map<int, int>> memo(n);
        for (int r = 0; r < n; r++)
        {
            for (int l = 0; l < r; l++)
            {
                int diff = nums[r] - nums[l];
                if (memo[l].find(diff) != memo[l].end())
                {
                    memo[r][diff] = memo[l][diff] + 1;
                }
                else
                {
                    memo[r][diff] = 2;
                }
                maxLen = max(memo[r][diff], maxLen);
            }
        }
        return maxLen;
    }
};

File: q1027/q1027.java
class Solution {
    public int longestArithSeqLength(int[] nums) {
        HashMap<Integer, Integer>[] memo = new HashMap[nums.length];
        // Arrays.fill(memo, new HashMap<Integer, Integer>());
        int maxLen = 0;
        for (int r = 0; r < nums.length; r++) {
            memo[r] = new HashMap<Integer, Integer>();
            for (int l = 0; l < r; l++) {
                int diff = nums[l] - nums[r];
                int prevMaxLen = memo[l].getOrDefault(diff, 1);
                memo[r].put(diff, prevMaxLen + 1);
                maxLen = Math.max(maxLen, memo[r].get(diff));
            }
        }
        return maxLen;
    }
}